### PR TITLE
En 3649 too many open files

### PR DESCRIPTION
**Pull request by BeniaminDrasovean**
Added the option to specify in config the maximum files that a storage unit can open. This solves the "out of memory" and "too many open files" bugs.
Updated resourceMonitor.

### CODE REVIEW

File: core/statistics/resourceMonitor.go [hunk 1]
```diff
@@ -76,3 +97,13 @@ func (rm *ResourceMonitor) Close() error {
 	rm.file = nil
 	return err
 }
+
+func getCurrentProcess() (*process.Process, error) {
```
Comment (miiu96) [inline core/statistics/resourceMonitor.go]: This function is now in statistics/machine/common.go and you should move file common.go in statistics folder to can access function getCurrentProcess() also in machine package and statistics package.

File: core/statistics/resourceMonitor.go [hunk 2]
```diff
@@ -34,14 +35,34 @@ func (rm *ResourceMonitor) GenerateStatistics() string {
 	var memStats runtime.MemStats
 	runtime.ReadMemStats(&memStats)
 
-	return fmt.Sprintf("timestamp: %d, uptime: %v, num go: %d, go mem: %s, sys mem: %s, total mem: %s, num GC: %d\n",
+	fileDescriptors := int32(0)
+	numOpenFiles := 0
+	numConns := 0
+	proc, err := getCurrentProcess()
+	if err == nil {
```
Comment (miiu96) [inline core/statistics/resourceMonitor.go]: what you will you if err is not equal with nil ?

File: core/statistics/resourceMonitor.go [hunk 3]
```diff
@@ -34,14 +35,34 @@ func (rm *ResourceMonitor) GenerateStatistics() string {
...
+		fileDescriptors, _ = proc.NumFDs()
+		openFiles, err := proc.OpenFiles()
+		if err == nil {
+			numOpenFiles = len(openFiles)
```
Comment (camilbancioiu) [inline core/statistics/resourceMonitor.go]: Is the number of file descriptors currently in use different from the number of open files?

File: storage/leveldb/leveldb.go [hunk 4]
```diff
@@ -30,15 +30,16 @@ type DB struct {
 
 // NewDB is a constructor for the leveldb persister
 // It creates the files in the location given as parameter
-func NewDB(path string, batchDelaySeconds int, maxBatchSize int) (s *DB, err error) {
+func NewDB(path string, batchDelaySeconds int, maxBatchSize int, maxOpenFiles int) (s *DB, err error) {
 	err = os.MkdirAll(path, rwxOwner)
 	if err != nil {
 		return nil, err
 	}
```
Comment (camilbancioiu) [inline storage/leveldb/leveldb.go]: Should there be a verification against a minimum possible number for `maxOpenFiles`? Would `leveldb` work well when `OpenFilesCacheCapacity = 0`?

File: storage/leveldb/leveldbSerial.go [hunk 5]
```diff
@@ -28,15 +28,16 @@ type SerialDB struct {
 
 // NewSerialDB is a constructor for the leveldb persister
 // It creates the files in the location given as parameter
-func NewSerialDB(path string, batchDelaySeconds int, maxBatchSize int) (s *SerialDB, err error) {
+func NewSerialDB(path string, batchDelaySeconds int, maxBatchSize int, maxOpenFiles int) (s *SerialDB, err error) {
 	err = os.MkdirAll(path, rwxOwner)
 	if err != nil {
 		return nil, err
 	}
```
Comment (camilbancioiu) [inline storage/leveldb/leveldbSerial.go]: Should there be a verification against a minimum possible number for `maxOpenFiles`? Would `leveldb` work well when `OpenFilesCacheCapacity = 0`?

File: core/statistics/resourceMonitor.go (hunk 1 above)
Comment (BeniaminDrasovean) [reply core/statistics/resourceMonitor.go]: Done. If I move common.go in statistics, I won't have access to the function from the machine subpackage, so I exported the func.

File: core/statistics/resourceMonitor.go (hunk 3 above)
Comment (BeniaminDrasovean) [reply core/statistics/resourceMonitor.go]: No, they differ. At least on linux.

File: storage/leveldb/leveldb.go (hunk 4 above)
Comment (BeniaminDrasovean) [reply storage/leveldb/leveldb.go]: Will test to see the appropriate boundaries.

File: storage/leveldb/leveldbSerial.go (hunk 5 above)
Comment (BeniaminDrasovean) [reply storage/leveldb/leveldbSerial.go]: Will test to see the appropriate boundaries.

Comment (BeniaminDrasovean) [reply storage/leveldb/leveldbSerial.go]: Done.

File: storage/leveldb/leveldb.go (hunk 4 above)
Comment (BeniaminDrasovean) [reply storage/leveldb/leveldb.go]: Done.

**Pull request merged**